packages/opentelemetry-scope-zone-peer-dep/src/ZoneScopeManager.ts

```diff
@@ -0,0 +1,511 @@
+import { isListenerObject } from './util';
+
+/**
+ * Zone Scope Manager for managing the state in web
+ * it fully supports the async calls
+ */
+export class ZoneScopeManager implements ScopeManager {
+  /**
+   * whether the scope manager is enabled or not
+   */
+  private _enabled = false;
+
+  /**
+   * Keeps the the information whether to prevent restoring the scope after zone onInvoke
+   */
+  private _preventScopeRestoring: { [key: string]: boolean } = {};
+
+  /**
+   * Keeps the reference to the scopes per zone
+   */
+  private _scopes: { [key: string]: unknown[] } = {};
+
+  /**
+   * Keeps the reference to the scopes with original event targets
+   */
+  private _scopesWithEventTargets: ScopeWithEventTargets[] = [];
+
+  /**
+   * Helps to create a unique name for the zones - part of zone name
+   */
+  private _zoneCounter = 0;
+
+  constructor() {}
+
+  /**
+   * Returns the active scope from certain zone name
+   * @param activeZoneName
+   */
+  private _activeScopeFromZoneName(
+    activeZoneName: string | undefined
+  ): unknown | undefined {
+    if (activeZoneName) {
+      const scopes = this._scopes[activeZoneName];
+      if (scopes) {
+        return scopes[scopes.length - 1];
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * @param target Function to be executed within the scope
+   * @param scope
+   */
+  private _bindFunction<T extends Function>(target: T, scope?: unknown): T {
+    const manager = this;
+    const contextWrapper = function(this: any, ...args: unknown[]) {
+      return manager.with(scope, () => target.apply(this || scope, args));
+    };
+    Object.defineProperty(contextWrapper, 'length', {
+      enumerable: false,
+      configurable: true,
+      writable: false,
+      value: target.length,
+    });
+    return (contextWrapper as unknown) as T;
+  }
+
+  /**
+   * @param obj target object on which the listeners will be patched
+   * @param scope
+   */
+  private _bindListener<T>(obj: T, scope?: unknown): T {
+    const target = (obj as unknown) as TargetWithEvents;
+    if (target.__ot_listeners !== undefined) {
+      return obj;
+    }
+    target.__ot_listeners = {};
+
+    if (typeof target.addEventListener === 'function') {
+      target.addEventListener = this._patchAddEventListener(
+        target,
+        target.addEventListener,
+        scope
+      );
+    }
+
+    if (typeof target.removeEventListener === 'function') {
+      target.removeEventListener = this._patchRemoveEventListener(
+        target,
+        target.removeEventListener,
+        scope
+      );
+    }
+
+    return obj;
+  }
+
+  /**
+   * Removes references to the targets that were bind to certain scope / span
+   * @param scope
+   */
+  private _cleanEventTargetsFromScope(scope: any, target?: TargetWithEvents) {
+    // console.log('_cleanEventTargetsFromScope1', this._scopesWithEventTargets.length, scope && scope.name);
+    for (let i = this._scopesWithEventTargets.length - 1; i >= 0; i--) {
```

Comment:
Could you use `for (const .. of` here?

Reply:
I can't as I need an index here to remove element from array

Reply:
you should be able to do `for (const [i, scope] of this._scopesWithEventTargets.entries()) {`

Reply:
@dyladan you will still go from first instead of last element and such structure is much slower then standard loop